pkg/pubsub/system.go

```diff
@@ -0,0 +1,82 @@
+package pubsub
+
+import (
+	"context"
+	"sync"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/pubsub/connection"
+	"github.com/open-policy-agent/gatekeeper/pkg/pubsub/provider"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var log = logf.Log.WithName("pubsub-system")
+
+type System struct {
+	mux         sync.RWMutex
+	connections map[string]connection.Connection
+	providers   map[string]string
+}
+
+func NewSystem() *System {
+	return &System{}
+}
+
+func (p *System) Publish(ctx context.Context, connection string, topic string, msg interface{}) error {
+	p.mux.RLock()
+	defer p.mux.RUnlock()
+	if len(p.connections) > 0 {
+		if c, ok := p.connections[connection]; ok {
+			return c.Publish(context.Background(), msg, topic)
+		}
+		log.Info("Connection is not initialized", "name", connection)
+	} else {
+		log.Info("No pub sub tools are enabled")
+	}
+	return nil
+}
+
+func (p *System) UpsertConnection(ctx context.Context, newConnection provider.InitiateConnection, config interface{}, name string, provider string) error {
+	p.mux.Lock()
+	defer p.mux.Unlock()
+	if _, ok := p.connections[name]; ok {
+		log.Info("this is an update request")
+		if p.providers[name] != provider {
+			// if provider is updated then close connection for old provider and create connection with updated provider
+			err := p.CloseConnection(name)
+			if err != nil {
+				return err
+			}
+		} else {
+			return p.connections[name].UpdateConnection(ctx, config)
```

Comment:
I like that the connection object is responsible for interpreting the meaning of an update